### PR TITLE
Fusion: Resolution validator

### PR DESCRIPTION
**Pull request by EmberLightVFX**
## Changelog Description
Added a resolution validator.
The code is from my old PR (https://github.com/ynput/OpenPype/pull/4921) that I closed because the PR also contained a frame range validator that no longer is needed.

## Additional info
The validator is optional as you might want/need to render out different resolutions than what's set for the asset.

## Testing notes:
1. Make a BG node with a different resolution than your asset
2. Create a new saver and pipe in the BG
3. Validate and see an error pop up

### CODE REVIEW

File: openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    order = pyblish.api.ValidatorOrder
+    label = "Validate Saver Resolution"
+    families = ["render"]
+    hosts = ["fusion"]
+    optional = True
+    actions = [SelectInvalidAction]
+
+    @classmethod
+    def get_invalid(cls, instance, wrong_resolution=None):
+        """Validate if the saver rescive the expected resolution"""
```
Comment (BigRoy) [inline openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py:28]: rescive?

```suggestion
        """Validate if the saver receives the expected resolution"""
```

File: openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        if wrong_resolution is None:
+            wrong_resolution = []
+
+        saver = instance[0]
+        firstFrame = saver.GetAttrs("TOOLNT_Region_Start")[1]
+        comp = get_current_comp()
+
+        # If the current saver hasn't bin rendered its input resolution
+        # hasn't bin saved. To combat this, add an expression in
+        # the comments field to read the resolution
```
Comment (BigRoy) [inline openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py:38]:
```suggestion
        # If the current saver hasn't been rendered its input resolution
        # hasn't been saved. To combat this, add an expression in
        # the comments field to read the resolution
```

File: openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    optional = True
+    actions = [SelectInvalidAction]
+
+    @classmethod
+    def get_invalid(cls, instance, wrong_resolution=None):
+        """Validate if the saver rescive the expected resolution"""
+        if wrong_resolution is None:
+            wrong_resolution = []
+
+        saver = instance[0]
```
Comment (jakubjezek001) [inline openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py:32]:
```suggestion
        saver = instance.data["tool"]
```

Comment (jakubjezek001) [reply openpype/hosts/fusion/plugins/publish/validator_saver_resolution.py:32]: There is not a tool in the list position anymore.

**Review by jakubjezek001 — CHANGES_REQUESTED** [review]
Thanks for contribution @EmberLightVFX. Please address the fixes and then it is good to be merged.

**Review by BigRoy — APPROVED** [review]
Works as intended.

image

Only doubt I have is maybe the label on the validator - would it make sense to label that "Validate Asset Resolution" instead of "Validate Saver Resolution" since it might make more sense to the artist what it is being validated against?

image

File: openpype/hosts/fusion/plugins/publish/validate_saver_resolution.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        return width, height
+
+
+class ValidateSaverResolution(
+    pyblish.api.InstancePlugin, OptionalPyblishPluginMixin
+):
+    """Validate that the saver input resolution matches the asset resolution"""
+
+    order = pyblish.api.ValidatorOrder
+    label = "Validate Saver Resolution"
```
Comment (BigRoy) [inline openpype/hosts/fusion/plugins/publish/validate_saver_resolution.py:66]:
```suggestion
    label = "Validate Asset Resolution"
```

**Review by jakubjezek001 — APPROVED** [review]
LGTM. Thank you!

**Comment by BigRoy** [comment]
@jakubjezek001 ⚠️ This PR is lacking matching settings for AYON so a fix PR to create the missing Fusion settings for AYON should be generated.

**Pull request merged**